Accurate Bresenham raytracing implementation and clearing for corner cases

AlexReimann opened the pull request:
Changes / fixes clearing with Bresenham raytracing using a more accurate version
Adds parameters for accuracy settings
Adds parameter for including corner cases into clearing

See also https://github.com/ros-planning/navigation/issues/267

Implementation based on: http://graphics.idav.ucdavis.edu/education/GraphicsNotes/CAGDNotes/Bresenhams-Algorithm.pdf

AlexReimann committed:
Changes / fixes clearing with Bresenham raytracing using a more accurate version
Adds parameters for accuracy settings
Adds parameter for including corner cases into clearing

mikeferguson commented on costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -62,6 +62,14 @@ void VoxelLayer::onInitialize()
 
   clearing_endpoints_pub_ = private_nh.advertise<sensor_msgs::PointCloud>( "clearing_endpoints", 1 );
+
+  combination_method_ = 1;
```
should this actually be here?

mikeferguson commented on voxel_grid/include/voxel_grid/voxel_grid.h:
```diff
@@ -86,6 +86,27 @@ class VoxelGrid
+   *
+   * Number of bits should be less than:
+   * (bits representing the max of an integer) minus
+   * (bits needed to represent the * maximum raytrace distance (= maximum sensor range) in cells)
```
Should that star (*) be in the middle of the line (before maximum raytrace)?

mikeferguson commented on voxel_grid/include/voxel_grid/voxel_grid.h:
```diff
@@ -226,15 +248,20 @@ class VoxelGrid
-    int dx = int(x1) - int(x0);
-    int dy = int(y1) - int(y0);
-    int dz = int(z1) - int(z0);
+    double dx = x1 - x0;
```
do you have any numbers on the performance penalty of switching all this from int to double?

AlexReimann replied:
Nope, but it still uses integers in the actual ray tracing which is the most work.

mikeferguson commented on voxel_grid/include/voxel_grid/voxel_grid.h:
```diff
@@ -226,15 +248,20 @@ class VoxelGrid
-    unsigned int abs_dz = abs(dz);
+    if(abs_dx > INT_MAX || abs_dy > INT_MAX || abs_dz > INT_MAX)
+    {
+      ROS_ERROR("Voxel grid: Accuracy multiplier is set too high. Possible integer overflow while clearing.");
```
this should probably be a ROS_ERROR_THROTTLE, else too high of a setting is likely to tank your computer with messages.

mikeferguson commented on voxel_grid/include/voxel_grid/voxel_grid.h:
```diff
@@ -243,38 +270,72 @@ class VoxelGrid
-      int error_z = abs_dx / 2;
+      int error_y = (int)(abs_dy * x_lost_rounding - abs_dx * y_lost_rounding);
+      int error_z = (int)(abs_dz * x_lost_rounding - abs_dx * z_lost_rounding);
+      int number_of_cells = abs(int(x0) - int(x1));
```
this is maybe a little nitpicky -- but this number is not actually number_of_cells yet (until multiplied by scale). I would suggest fixing these lines (3 of them, plus 6 calls to bresenham functions) to do the multiplication by scale here (and also make them unsigned int to begin with). Long term, this should avoid people potentially changing one line below and not the others.

AlexReimann replied:
It is used with and without multiplying by scale, but after looking at it again it doesn't really make sense:
It is later used with min() and scale has to be between 0 and 1. So I'll just multiply it always with scale.
(The whole thing should be already done way before here imho, but maybe in a later patch)

mikeferguson commented:
This looks pretty good -- I would suggest adding a link to that PDF somewhere in the code for any future developer/maintainer.

This is a pretty big change to a core piece of the navigation stack, so we will leave this PR open at least a week so that others have time to test it, and comment if they have any issues.

AlexReimann commented:
Known bug: Clearing corner cases can lead to underflow / overflow of the clearing which then leads to actually marking
So on hold till I find a fix which does not take up too much computation

AlexReimann commented:
The issue is the accuracy and it only happens when the end point is directly on the z- axis border of voxel grid cells. Also occurence depends on the accuracy.
One fix would be to check if the end point is on the border and add / subtract a little offset, might be computational "heavy".

The other one would be to remove corner clearing for the last cleared cell. Basically it does this right now: https://docs.google.com/drawings/d/1YeeKCFF7LxVRLOkIBawkwEzePvb2JeTrjfjLt8b18Bk/edit?usp=sharing

So if we skip the clearing of additional corner cells after clearing the last cell it fixes it. But might under clear? I will do some more testing on that again.. if it works we might want to go with that.
Opinions on that?

Third option would be to check for underflow /overflow and don't clear in this case. But this is even more computational heavy with the current implementation than the first proposed fix.

I have another pull request incoming after this which changes the way updating is handled. It makes the whole thing around 30% faster (with a kinect) and also allows to visualize which cells are cleared (which is really nice for debugging). It would also allow an underflow detection without slowing it down much (maybe 1% ?).
Non-optimized version, which is actually slower but has the visualisation, is here https://github.com/AlexReimann/navigation/commit/e20243d643d0d497352a176797efb1b1945f3096

Commit:
Adds max_raytracing_range variable to obstacle layer which is needed for cached clearing of voxel grid
Moves grid markers and clearers into seperated files and does some refactoring on them
Adds parameter to choose between cached clearing and normal / old style clearing
Adds publisher for cleared voxels when using the cached clearing
Cached clearing now keeps track of which cells in the costmap are cleared for future use

AlexReimann commented:
I introduced a cached updating scheme of the costmap which fetches overflows / underflows in case of corner case clearing.
It also makes the code faster than the original implementation in the case that number of voxels << number of cleared voxels (e. g. 3d sensor where most voxels get cleared a few hundred times depending on the resolution).
When using corner case clearing it now runs at around the same speed as the original implementation or faster (depends on the length of rays).
Also adds visualization of the cleared voxels.
This needs some serious testing.

AlexReimann commented:
Should I re-open the pull request in indigo-devel?

mikeferguson commented:
Yes please.

AlexReimann commented:
Reopend on indigo-devel: #316
